// @datadog/datadog-ci/dist/commands/trace/api.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiConstructor = exports.reportCustomSpan = void 0;
const git_1 = require("../../helpers/git");
const user_provided_git_1 = require("../../helpers/user-provided-git");
const utils_1 = require("../../helpers/utils");
// Dependency follows-redirects sets a default maxBodyLength of 10 MB https://github.com/follow-redirects/follow-redirects/blob/b774a77e582b97174813b3eaeb86931becba69db/index.js#L391
// We don't want any hard limit enforced by the CLI, the backend will enforce a max size by returning 413 errors.
const maxBodyLength = Infinity;
const reportCustomSpan = (request) => (customSpan, provider) => __awaiter(void 0, void 0, void 0, function* () {
    const gitSpanTags = yield git_1.getGitMetadata();
    const userGitSpanTags = user_provided_git_1.getUserGitSpanTags();
    return request({
        data: Object.assign(Object.assign({}, customSpan), { tags: Object.assign(Object.assign(Object.assign({}, gitSpanTags), userGitSpanTags), customSpan.tags) }),
        headers: {
            'X-Datadog-CI-Custom-Event': provider,
        },
        maxBodyLength,
        method: 'POST',
        url: 'api/v2/webhook',
    });
});
exports.reportCustomSpan = reportCustomSpan;
const apiConstructor = (baseUrl, apiKey) => {
    const requestIntake = utils_1.getRequestBuilder({ baseUrl, apiKey });
    return {
        reportCustomSpan: exports.reportCustomSpan(requestIntake),
    };
};
exports.apiConstructor = apiConstructor;
