// @datadog/datadog-ci/dist/commands/git-metadata/__tests__/git.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const simpleGit = __importStar(require("simple-git"));
const git_1 = require("../git");
const createMockSimpleGit = (conf) => ({
    getRemotes: (_) => __awaiter(void 0, void 0, void 0, function* () {
        if (conf.remotes === undefined) {
            throw Error('Unexpected call to getRemotes');
        }
        return conf.remotes;
    }),
    raw: (command) => __awaiter(void 0, void 0, void 0, function* () {
        if (command === 'ls-files' && conf.trackedFiles !== undefined) {
            return conf.trackedFiles.join('\n') + '\n';
        }
        throw Error(`Unexpected call to raw(${command})`);
    }),
    revparse: (_) => __awaiter(void 0, void 0, void 0, function* () {
        if (conf.hash === undefined) {
            throw Error('Unexpected call to revparse');
        }
        return conf.hash;
    }),
});
describe('git', () => {
    describe('gitRemote', () => {
        test('should choose the remote named origin', () => __awaiter(void 0, void 0, void 0, function* () {
            const mock = createMockSimpleGit({
                remotes: [
                    { name: 'first', refs: { push: 'remote1' } },
                    { name: 'origin', refs: { push: 'remote2' } },
                ],
            });
            const remote = yield git_1.gitRemote(mock);
            expect(remote).toBe('remote2');
        }));
        test('should choose the first remote', () => __awaiter(void 0, void 0, void 0, function* () {
            const mock = createMockSimpleGit({
                remotes: [
                    { name: 'first', refs: { push: 'remote1' } },
                    { name: 'second', refs: { push: 'remote2' } },
                ],
            });
            const remote = yield git_1.gitRemote(mock);
            expect(remote).toBe('remote1');
        }));
    });
    describe('stripCredentials: git protocol', () => {
        test('should return the same value', () => {
            const input = 'git@github.com:user/project.git';
            expect(git_1.stripCredentials(input)).toBe(input);
        });
    });
    describe('stripCredentials: nothing to remove', () => {
        test('should return the same value', () => {
            const input = 'https://gitlab.com/user/project.git';
            expect(git_1.stripCredentials(input)).toBe(input);
        });
    });
    describe('stripCredentials: user:pwd', () => {
        test('should return without credentials', () => {
            const input = 'https://token:[MASKED]@gitlab.com/user/project.git';
            expect(git_1.stripCredentials(input)).toBe('https://gitlab.com/user/project.git');
        });
    });
    describe('stripCredentials: token', () => {
        test('should return without credentials', () => {
            const input = 'https://token@gitlab.com/user/project.git';
            expect(git_1.stripCredentials(input)).toBe('https://gitlab.com/user/project.git');
        });
    });
    describe('getCommitInfo', () => {
        test('should return commit info from simple git', () => __awaiter(void 0, void 0, void 0, function* () {
            const mock = createMockSimpleGit({
                hash: 'abcd',
                remotes: [{ name: 'first', refs: { push: 'https://git-repo' } }],
                trackedFiles: ['myfile.js'],
            });
            const commitInfo = yield git_1.getCommitInfo(mock);
            expect(commitInfo).toBeDefined();
            expect(commitInfo.hash).toBe('abcd');
            expect(commitInfo.trackedFiles).toStrictEqual(['myfile.js']);
            expect(commitInfo.remote).toBe('https://git-repo/');
        }));
        test('should return commit info with overridden repo name', () => __awaiter(void 0, void 0, void 0, function* () {
            const mock = createMockSimpleGit({
                hash: 'abcd',
                trackedFiles: ['myfile.js'],
            });
            const commitInfo = yield git_1.getCommitInfo(mock, 'https://overridden');
            expect(commitInfo).toBeDefined();
            expect(commitInfo.hash).toBe('abcd');
            expect(commitInfo.trackedFiles).toStrictEqual(['myfile.js']);
            expect(commitInfo.remote).toBe('https://overridden');
        }));
    });
    describe('newSimpleGit', () => {
        test('should throw an error if git is not installed', () => __awaiter(void 0, void 0, void 0, function* () {
            jest.spyOn(simpleGit, 'gitP').mockImplementation(() => {
                throw Error('gitp error');
            });
            yield expect(git_1.newSimpleGit()).rejects.toThrow('gitp error');
        }));
        test('should throw an error if revparse throws an error', () => __awaiter(void 0, void 0, void 0, function* () {
            const mock = createMockSimpleGit({});
            jest.spyOn(simpleGit, 'gitP').mockReturnValue(mock);
            jest.spyOn(mock, 'revparse').mockImplementation(() => __awaiter(void 0, void 0, void 0, function* () {
                throw Error('revparse error');
            }));
            yield expect(git_1.newSimpleGit()).rejects.toThrow('revparse error');
        }));
        test('should not throw any errors', () => __awaiter(void 0, void 0, void 0, function* () {
            const mock = createMockSimpleGit({});
            jest.spyOn(simpleGit, 'gitP').mockReturnValue(mock);
            jest.spyOn(mock, 'revparse').mockResolvedValue('1234');
            yield expect(git_1.newSimpleGit()).resolves.not.toThrow();
        }));
    });
});
