// @datadog/datadog-ci/dist/commands/junit/renderer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderCommandInfo = exports.renderUpload = exports.renderDryRunUpload = exports.renderSuccessfulCommand = exports.renderRetriedUpload = exports.renderFailedUpload = exports.renderInvalidFile = void 0;
const chalk_1 = __importDefault(require("chalk"));
const path_1 = __importDefault(require("path"));
const ICONS = {
    FAILED: '❌',
    SUCCESS: '✅',
    WARNING: '⚠️',
};
const renderInvalidFile = (xmlPath, errorMessage) => {
    const jUnitXMLPath = `[${chalk_1.default.bold.dim(xmlPath)}]`;
    return chalk_1.default.red(`${ICONS.FAILED} Invalid jUnitXML file ${jUnitXMLPath}: ${errorMessage}\n`);
};
exports.renderInvalidFile = renderInvalidFile;
const renderFailedUpload = (payload, errorMessage) => {
    const jUnitXMLPath = `[${chalk_1.default.bold.dim(payload.xmlPath)}]`;
    return chalk_1.default.red(`${ICONS.FAILED} Failed upload jUnitXML for ${jUnitXMLPath}: ${errorMessage}\n`);
};
exports.renderFailedUpload = renderFailedUpload;
const renderRetriedUpload = (payload, errorMessage, attempt) => {
    const jUnitXMLPath = `[${chalk_1.default.bold.dim(payload.xmlPath)}]`;
    return chalk_1.default.yellow(`[attempt ${attempt}] Retrying jUnitXML upload ${jUnitXMLPath}: ${errorMessage}\n`);
};
exports.renderRetriedUpload = renderRetriedUpload;
const renderSuccessfulCommand = (fileCount, duration) => chalk_1.default.green(`${ICONS.SUCCESS} Uploaded ${fileCount} files in ${duration} seconds.\n`);
exports.renderSuccessfulCommand = renderSuccessfulCommand;
const renderDryRunUpload = (payload) => `[DRYRUN] ${exports.renderUpload(payload)}`;
exports.renderDryRunUpload = renderDryRunUpload;
const renderUpload = (payload) => `Uploading jUnit XML test report file in ${payload.xmlPath}\n`;
exports.renderUpload = renderUpload;
const renderCommandInfo = (basePaths, service, concurrency, dryRun) => {
    let fullStr = '';
    if (dryRun) {
        fullStr += chalk_1.default.yellow(`${ICONS.WARNING} DRY-RUN MODE ENABLED. WILL NOT UPLOAD JUNIT XML\n`);
    }
    fullStr += chalk_1.default.green(`Starting upload with concurrency ${concurrency}. \n`);
    if (basePaths.length === 1 && !!path_1.default.extname(basePaths[0])) {
        fullStr += chalk_1.default.green(`Will upload jUnit XML file ${basePaths[0]}\n`);
    }
    else {
        fullStr += chalk_1.default.green(`Will look for jUnit XML files in ${basePaths.join(', ')}\n`);
    }
    fullStr += chalk_1.default.green(`service: ${service}\n`);
    return fullStr;
};
exports.renderCommandInfo = renderCommandInfo;
