// @datadog/datadog-ci/dist/commands/metric/metric.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetricCommand = exports.parseMetrics = void 0;
const chalk_1 = __importDefault(require("chalk"));
const clipanion_1 = require("clipanion");
const ci_1 = require("../../helpers/ci");
const retry_1 = require("../../helpers/retry");
const utils_1 = require("../../helpers/utils");
const parseMetrics = (metrics) => metrics.reduce((acc, keyValue) => {
    if (!keyValue.includes(':')) {
        throw new Error(`invalid metrics key value pair "${keyValue}"`);
    }
    const [key, value] = keyValue.split(':', 2);
    const floatVal = parseFloat(value);
    if (isNaN(floatVal)) {
        throw new Error('value is not numeric');
    }
    return Object.assign(Object.assign({}, acc), { [key]: floatVal });
}, {});
exports.parseMetrics = parseMetrics;
class MetricCommand extends clipanion_1.Command {
    constructor() {
        super(...arguments);
        this.config = {
            apiKey: process.env.DATADOG_API_KEY || process.env.DD_API_KEY,
        };
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.level !== 'pipeline' && this.level !== 'job') {
                this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} Level must be one of [pipeline, job]\n`);
                return 1;
            }
            if (!this.metrics || this.metrics.length === 0) {
                this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} --metrics is required\n`);
                return 1;
            }
            try {
                const metrics = exports.parseMetrics(this.metrics);
                const { provider, ciEnv } = ci_1.getCIEnv();
                // For GitHub only the pipeline level is supported as there is no way to identify the job from the runner.
                if (provider === 'github' && this.level === 'job') {
                    this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} Cannot use level "job" for GitHub Actions.`);
                    return 1;
                }
                const exitStatus = yield this.sendMetrics(ciEnv, this.level === 'pipeline' ? 0 : 1, provider, metrics);
                if (exitStatus !== 0 && this.noFail) {
                    this.context.stderr.write(`${chalk_1.default.yellow.bold('[WARNING]')} sending metrics failed but continuing due to --no-fail\n`);
                    return 0;
                }
                else if (exitStatus === 0) {
                    this.context.stdout.write('Metrics sent\n');
                }
                return exitStatus;
            }
            catch (error) {
                this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} ${error.message}\n`);
                return 1;
            }
        });
    }
    sendMetrics(ciEnv, level, provider, metrics) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.config.apiKey) {
                this.context.stdout.write(`Neither ${chalk_1.default.red.bold('DATADOG_API_KEY')} nor ${chalk_1.default.red.bold('DD_API_KEY')} is in your environment.\n`);
                throw new Error('API key is missing');
            }
            const site = process.env.DATADOG_SITE || process.env.DD_SITE || 'datadoghq.com';
            const baseAPIURL = `https://${utils_1.getApiHostForSite(site)}`;
            const request = utils_1.getRequestBuilder({ baseUrl: baseAPIURL, apiKey: this.config.apiKey });
            const doRequest = () => request({
                data: {
                    data: {
                        attributes: {
                            ci_env: ciEnv,
                            ci_level: level,
                            metrics,
                            provider,
                        },
                        type: 'ci_custom_metric',
                    },
                },
                method: 'post',
                url: 'api/v2/ci/pipeline/metrics',
            });
            try {
                yield retry_1.retryRequest(doRequest, {
                    onRetry: (e, attempt) => {
                        this.context.stderr.write(chalk_1.default.yellow(`[attempt ${attempt}] Could not send metrics. Retrying...: ${e.message}\n`));
                    },
                    retries: 5,
                });
            }
            catch (error) {
                this.context.stderr.write(`${chalk_1.default.red.bold('[ERROR]')} Could not send metrics: ${error.message}\n`);
                return 1;
            }
            return 0;
        });
    }
}
exports.MetricCommand = MetricCommand;
MetricCommand.usage = clipanion_1.Command.Usage({
    description: 'Add metrics to a CI Pipeline trace pipeline or job span in Datadog.',
    details: `
            This command when run from a supported CI provider sends an arbitrary set of key:value
            numeric tags to Datadog to include in the CI Visibility traces.
    `,
    examples: [
        ['Add a binary size to the current pipeline', 'datadog-ci metric --level pipeline --tags binary.size:500'],
        ['Tag the current CI job with a command runtime', 'datadog-ci metric --level job --tags command.runtime:67.1'],
    ],
});
MetricCommand.addPath('metric');
MetricCommand.addOption('noFail', clipanion_1.Command.Boolean('--no-fail'));
MetricCommand.addOption('metrics', clipanion_1.Command.Array('--metrics'));
MetricCommand.addOption('level', clipanion_1.Command.String('--level'));
